File: Phase_3/Code/Transaction.cpp
#include "Transaction.h"

void Transaction::Withdrawal(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin && (len == 2 || len == 3))
	{
		//variable for the current user
		BankAccount* User;
		int argindex = 0;

		//check for admin access
		if(admin)
		{
			//get currect bankaccount if null no account name exits
			User = GetBankAccount(args[argindex++], cmd);
			if(User == NULL)
				return;
		}
		else
		{
			//if standard login, use current acccount
			User = currentAccount;
		}

		//check for inactive
		if(!User->IsActive())
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
			return;
		}

		//check if the account number is ok, by get next argument
		int accountNumber = Formatting::StringToInt(args[argindex++]);
		if(User->GetNumber() != accountNumber)
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
			return;
		}

		//check if amount if correct, getting the last argument
		float amount = Formatting::StringToFloat(args[argindex]);
		if(admin)
		{
			//check if amount is between 0 < amount < 99999
			if(amount < 0.0f || amount > 99999.99f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;
			}
		}
		else
		{
			//check if amount is between 0 < amount < 500
			if(amount < 0.0f || amount > 500.00f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;
			}

			//check if amount has already ben take out
			if(withAmount >= 500.0f || withAmount + amount > 500.0f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;	
			}
			withAmount += amount;
		}

		//check if amount is a withdrawable amount, must be a factor of 5
		float value = ((int)amount) - amount;
		if(value != 0.0f || (((int)amount) % 5) != 0)
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
			return;
		}

		//check for if trans action charge is applied
		float dec = GetTransactionCharge(User);
		

		//remove money if there is enough
		float rem = User->GetBalance() - amount - dec;
		if(rem >= 0.0f)
		{
			//enougn money
			User->SetBalance(rem);
			UpdateTransactions(OutFormat::MakeOutput(cmd, User->GetName(), User->GetNumber(), amount, "  "));
		}
		else 
		{
			//not enough money
			cout << Formatting::CommandToString(cmd) <<": Invalid Funds" << endl;
			return;
		}
	}
	else
	{
		//if not logged in
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Transfer(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin && (len == 4 || len == 3))
	{
	
		//variable for the current user
		BankAccount* User;
		BankAccount* ToUser;
		int argindex = 0;

		//check for admin access
		if(admin)
		{
			//get currect bankaccount if null no account name exits
			User = GetBankAccount(args[argindex++], cmd);
			if(User == NULL)
				return;
		}
		else
		{
			//if standard login, use current acccount
			User = currentAccount;
		}

		//check for inactive
		if(!User->IsActive())
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
			return;
		}

		//check if the account number is ok
		int accountNumber = Formatting::StringToInt(args[argindex++]);
		if(User->GetNumber() != accountNumber)
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
			return;
		}

		//get second account
		int secondAccountNumber = Formatting::StringToInt(args[argindex++]);
		ToUser = SearchNumber(secondAccountNumber);
		if(ToUser == NULL)
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
			return;
		}

		//check for inactive of second account
		if(!User->IsActive())
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
			return;
		}

		//check if amount is correct
		float amount = Formatting::StringToFloat(args[argindex]);
		if(admin)
		{
			if(amount <= 0.0f || amount > 99999.99f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;
			}
		}
		else
		{
			if(amount <= 0.0f || amount > 1000.00f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;
			}


			//check if amount has already ben take out
			if(transferAmount >= 1000.0f || transferAmount + amount > 1000.0f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;	
			}
			transferAmount += amount;
		}

		//check for if trans action charge is applied
		float dec = GetTransactionCharge(User);
		
		//remove money if there is enough
		float rem = User->GetBalance() - amount - dec;
		if(rem >= 0.0f)
		{
			//enougn money
			User->SetBalance(rem);
			UpdateTransactions(OutFormat::MakeOutput(cmd, User->GetName(), User->GetNumber(), amount, "  "));
			UpdateTransactions(OutFormat::MakeOutput(cmd, ToUser->GetName(), ToUser->GetNumber(), amount, "  "));
		}
		else 
		{
			//not enough money
			cout << Formatting::CommandToString(cmd) <<": Invalid Funds" << endl;
			return;
		}
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Paybill(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin && (len == 4 || len == 3))
	{
		BankAccount* User;
		int argindex = 0;
		//check for admin access
		if(admin)
		{
			//get currect bankaccount if null no account name exits
			User = GetBankAccount(args[argindex++], cmd);
			if(User == NULL)
				return;
		}
		else
		{
			//if standard login, use current acccount
			User = currentAccount;
		}

		//check for inactive
		if(!User->IsActive())
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
			return;
		}

		//check if the account number is ok
		int accountNumber = Formatting::StringToInt(args[argindex++]);
		if(User->GetNumber() != accountNumber)
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
			return;
		}

		//extract company and check if it is a good company name
		string Company = args[argindex++];
		if(Company.compare("EC") != 0 && Company.compare("CQ") != 0 && Company.compare("TV") != 0)
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Company Name" << endl;
			return;
		}

		//check if amount if correct
		float amount = Formatting::StringToFloat(args[argindex]);
		if(admin)
		{
			if(amount < 0.0f || amount > 99999.99f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;
			}
		}
		else
		{
			if(amount < 0.0f || amount > 2000.00f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;
			}

			//check if amount has already ben take out
			if(payAmount >= 2000.0f || payAmount + amount > 2000.0f)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
				return;	
			}
			payAmount += amount;
		}

		//check for if trans action charge is applied
		float dec = GetTransactionCharge(User);

		//remove money if there is enough
		float rem = User->GetBalance() - amount - dec;
		if(rem >= 0.0f)
		{
			//enougn money
			User->SetBalance(rem);
			UpdateTransactions(OutFormat::MakeOutput(cmd, User->GetName(), User->GetNumber(), amount, Company));
		}
		else 
		{
			//not enough money
			cout << Formatting::CommandToString(cmd) <<": Invalid Funds" << endl;
			return;
		}
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Deposit(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin && (len == 2 || len == 3))
	{
		BankAccount* User;
		int argindex = 0;
		//check for admin access
		if(admin)
		{

			//get currect bankaccount if null no account name exits
			User = GetBankAccount(args[argindex++], cmd);
			if(User == NULL)
				return;
		}
		else
		{
			//if standard login, use current acccount
			User = currentAccount;
		}

		//check for inactive
		if(!User->IsActive())
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
			return;
		}

		//check if the account number is ok
		int accountNumber = Formatting::StringToInt(args[argindex++]);
		if(User->GetNumber() != accountNumber)
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
			return;
		}

		//check if amount if correct
		float amount = Formatting::StringToFloat(args[argindex]);
		if(amount <= 0.0f || amount > 99999.99f)
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
			return;
		}


		//check for if trans action charge is applied
		float dec = GetTransactionCharge(User);

		//set the new balace and write the output
		User->SetBalance(User->GetBalance() - dec);
		UpdateTransactions(OutFormat::MakeOutput(cmd, User->GetName(), User->GetNumber(), amount, "  "));		
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Create(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin && len == 2)
	{
		if(admin)
		{	
			//admin only
			int argindex = 0;	
			string name = args[argindex++];
			
			//check if doesn't exits
			if(!CheckForBankAccount(name))
			{	
				//name must be less than 20, it shorteds it if it is above 20
				if(name.length() > 20)
				{
					name = name.substr(0, 20);
					cout << Formatting::CommandToString(cmd) <<": Shortened Character Amount" << endl;
				}
				
				//check if amount if correct
				float amount = Formatting::StringToFloat(args[argindex]);
				if(amount < 0.0f || amount > 99999.99f)
				{
					cout << Formatting::CommandToString(cmd) <<": Invalid Input" << endl;
					return;
				}

				//Search for avaliable number, if none avaliable, error out
				int accountNumber = SearchForNewNumber();
				if(accountNumber <= 0)
				{
					cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;
					return;
				}

				//write the output
				UpdateTransactions(OutFormat::MakeOutput(cmd, name, accountNumber, amount, "  "));
			}
			else
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Access" << endl;
				return;
			}
		}
		else
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Access" << endl;	
			return;
		}
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Delete(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin && len == 2)
	{
		if(admin)
		{
			BankAccount* User;
			int argindex = 0;

			//check for admin access
			User = GetBankAccount(args[argindex++], cmd);
			if(User == NULL)
				return;
			
			//check for inactive
			if(!User->IsActive())
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
				return;
			}

			//check if the account number is ok
			int accountNumber = Formatting::StringToInt(args[argindex++]);
			if(User->GetNumber() != accountNumber)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
				return;
			}

			//set the account to deleted and write the value out
			User->DeleteAccount();
			UpdateTransactions(OutFormat::MakeOutput(cmd, User->GetName(), User->GetNumber(), 0, "  "));
		}
		else
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Access" << endl;	
		}
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Disable(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin && len == 2)
	{
		if(admin)
		{
			//admin only
			BankAccount* User;
			int argindex = 0;

			//check for admin access
			User = GetBankAccount(args[argindex++], cmd);
			if(User == NULL)
				return;
			
			//check for inactive
			if(!User->IsActive())
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
				return;
			}

			//check if the account number is ok
			int accountNumber = Formatting::StringToInt(args[argindex++]);
			if(User->GetNumber() != accountNumber)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
				return;
			}

			//disable the user and write the command
			User->Disable();
			UpdateTransactions(OutFormat::MakeOutput(cmd, User->GetName(), User->GetNumber(), 0, "  "));
		}
		else
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Access" << endl;	
		}
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Changeplan(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin &&  len == 2 )
	{
		if(admin)
		{
			//admin only
			BankAccount* User;
			int argindex = 0;

			//check for admin access
			User = GetBankAccount(args[argindex++], cmd);
			if(User == NULL)
				return;
			
			//check for inactive
			if(!User->IsActive())
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
				return;
			}

			//check if the account number is ok
			int accountNumber = Formatting::StringToInt(args[argindex++]);
			if(User->GetNumber() != accountNumber)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
				return;
			}

			//write out the success of the command
			UpdateTransactions(OutFormat::MakeOutput(cmd, User->GetName(), User->GetNumber(), 0, User->ChangePlan()));
		}
		else
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Access" << endl;	
			return;
		}
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
		return;
	}
}

void Transaction::Enable(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin && len == 2)
	{
		if(admin)
		{
			//admin only
			BankAccount* User;
			int argindex = 0;

			//check for admin access
			User = GetBankAccount(args[argindex++], cmd);
			if(User == NULL)
				return;
			
			//check for inactive
			if(!User->IsActive())
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Access" << endl;	
				return;
			}

			//check if the account number is ok
			int accountNumber = Formatting::StringToInt(args[argindex++]);
			if(User->GetNumber() != accountNumber)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Number" << endl;
				return;
			}

			//enable the account and write the success of the command
			User->Enable();
			UpdateTransactions(OutFormat::MakeOutput(cmd, User->GetName(), User->GetNumber(), 0, "  "));

		}
		else
		{
			cout << Formatting::CommandToString(cmd) <<": Invalid Access" << endl;	
		}
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Login(string* args, int len, int cmd)
{
	//check for if logged in and if there is the right amount for argumentss
	if(logedin &&  len > 0)
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;
	}
	else if(args[0].compare("standard") == 0)
	{
		//for standard account login
		if(len > 2)
		{
			//make sure of only standard and a name, no extra data
			cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
		}
		else
		{
			//get the firs agument for the name
			string value = args[1];

			//check for account, if null don't log in
			currentAccount = GetBankAccount(value, cmd);
			if(currentAccount == NULL)
			{
				cout << Formatting::CommandToString(cmd) <<": Invalid Account Holder Name" << endl;			
				return;
			}

			//set login and admin, the write out the command
			withAmount = payAmount = transferAmount = 0.0f;
			logedin = true;
			admin = false;
			UpdateTransactions(OutFormat::MakeOutput(cmd, currentAccount->GetName(), currentAccount->GetNumber(), 0.0f, "S"));
		}
	}
	else if(args[0].compare("admin") == 0)
	{
		//for admin login
		if(len > 1)
		{
			//make sure only "admin" on the line, no extra data
			cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
		}
		else
		{
			//set account, login, admin, the write the command
			currentAccount = NULL;
			logedin = true;
			admin = true;
			UpdateTransactions(OutFormat::MakeOutput(cmd, "", 0, 0.0f, "A"));
		}
		
	}
	else
	{
		cout << Formatting::CommandToString(cmd) <<": Invalid Transaction" << endl;	
	}
}

void Transaction::Logout()
{
	//check for if logged in
	if(logedin)
	{
		//write out the respective logout
		if(admin)
		{
			UpdateTransactions(OutFormat::MakeOutput(0, "", 0, 0.0f, "A"));	
		}
		else
		{
			UpdateTransactions(OutFormat::MakeOutput(0, currentAccount->GetName(), currentAccount->GetNumber(), 0.0f, "S"));
		}
		//get ride of the current bankaccount
		currentAccount = NULL;
		
		//wrtie the currect transaction to the file
		ofstream out(filename.c_str(), fstream::app);
		out << transactionBuffer;
		out.close();

		//clear the transaciton buffer and logout
		transactionBuffer = "";
		logedin = false;
		admin = false;
	}
	else
	{
		cout << "logout: Invalid Transaction" << endl;
	}
}

File: Phase5/Transaction.java
class Transaction
{
	
	static public BankAccount Enable(BankAccount account)
	{
		account.setActive(true);
		return account;
	}

	static public BankAccount Disable(BankAccount account)
	{
		account.setActive(false);
		return account;
	}
	
	static public BankAccount Create(String name, int number, float amount)
	{
		return new BankAccount(number, name, true, false, amount, 0);
	}
	
	static public BankAccount Transfer(BankAccount account, float amount, boolean admin)
	{
		/*
		 * 1. check if bank accout is active. 
		 * 		- if not, throw an error message =
		 * 2. check if admin login or standard login
		 * 		- if standard, then transfer amount has to be between ( 0 > and < 1000)
		 * 3. check if student login, or non-student login. 
		 * 		- if non student charge --> $0.10.
		 * 		- if student charge --> $0.05
		 * 
		 */

		//the amount we are transfering from an account.
		float transferAmount = 0.0f;
		if(amount > 0){
			transferAmount = amount;
		}
		else{
			System.out.println("ERROR: transfer amount is less than $0. Transfer amount must be greater than $0.");
		}

		//if the account is Active
		if(account.isActive()){
			//check if its a non-student account
			if(!account.isStudent()){

				//An admin log in. 
				if(admin){

					//set transferAmount equal to the amount to transfer + the charges.  
					// transferAmount = amount + 0.10f;

					float newBalance = account.getBalance() - transferAmount ;
					if(newBalance < 0){
						System.out.println("ERROR");
					}
					else{
						account.setBalance(newBalance);
					}
				}

				//a standard login.
				else{
					//can't transfer more than $1000.00, on standard login.
					if(amount > 1000.00f){
						System.out.println("ERROR: Can not transfer more than $1000, on a standard login.");
					}

					//if amount < 1000.00, transfer the amount.
					else{
						//set transferAmount equal to the amount to transfer + the charges.  
						transferAmount = amount + 0.10f;

						float newBalance = account.getBalance() - transferAmount ;
						if(newBalance < 0){
							System.out.println("ERROR");
						}
						else{
							account.setBalance(newBalance);
						}
					}
				}
			}//end of !isStudent if statement.

			//student account
			else{
				//An admin log in. 
				if(admin){

					//set transferAmount equal to the amount to transfer + the charges.  
					// transferAmount = amount + 0.05f;

					float newBalance = account.getBalance() - transferAmount ;
					if(newBalance < 0){
						System.out.println("ERROR");
					}
					else{
						account.setBalance(newBalance);
					}
				}

				//a standard login.
				else{
					//can't transfer more than $1000.00, on standard login.
					if(amount > 1000.00f){
						System.out.println("ERROR: Can not transfer more than $1000, on a standard login.");
					}

					//if amount < 1000.00, transfer the amount.
					else{
						//set transferAmount equal to the amount to transfer + the charges.  
						transferAmount = amount + 0.05f;

						float newBalance = account.getBalance() - transferAmount ;
						if(newBalance < 0){
							System.out.println("ERROR");
						}
						else{
							account.setBalance(newBalance);
						}
					}
				}
			}
		}//end of isActive if statement.
		else{
			System.out.println("ERROR: Not an active account.");
		}
		return account;
	}
	
	static public BankAccount Withdrawal(BankAccount account, float amount, boolean admin)
	{
		/*
		 * withdrawal > 500 is not acceptable on standard login.
		 * withdrawal has to be multiple of 5.
		 */

		float withdrawalAmount = 0.0f;
		//check if amount is a multiple of 5.
		if(amount % 5 == 0){
			withdrawalAmount = amount;
		}else {
			System.out.println("ERROR: Amount is not a multiple of 5.");
		}

		//if the account is active
		if(account.isActive()){

			//non-student account
			if(!account.isStudent()){

				//admin access
				if(admin){
					// withdrawalAmount = amount + 0.10f;
					float newBalance = account.getBalance() - withdrawalAmount;

					if(newBalance  < 0){
						System.out.println("ERROR: Transaction cancled. After withdrawl the bank balance drops below 0.");
					}
					else{
						account.setBalance(newBalance);
					}

				}//end of acces type -> admin

				//standard access
				else{
					if(withdrawalAmount > 500){
						System.out.println("ERROR: Can not withdraw more than 500, on a standard login.");
					}
					else{

						withdrawalAmount = amount + 0.10f;
						float newBalance = account.getBalance() - withdrawalAmount;

						if(newBalance  < 0){
							System.out.println("ERROR: Transaction cancled. After withdrawl the bank balance drops below 0.");
						}
						else{
							account.setBalance(newBalance);
						}
					}

				}//end of access type -> std

			}//end of non-student account type

			//student account
			else{

				//admin access
				if(admin){
					// withdrawalAmount = amount + 0.05f;
					float newBalance = account.getBalance() - withdrawalAmount;

					if(newBalance  < 0){
						System.out.println("ERROR: Transaction cancled. After withdrawl the bank balance drops below 0.");
					}
					else{
						account.setBalance(newBalance);
					}

				}//end of acces type -> admin

				//standard access
				else{
					if(withdrawalAmount > 500){
						System.out.println("ERROR: Can not withdraw more than 500, on a standard login.");
					}
					else{

						withdrawalAmount = amount + 0.05f;
						float newBalance = account.getBalance() - withdrawalAmount;

						if(newBalance  < 0){
							System.out.println("ERROR: Transaction cancled. After withdrawl the bank balance drops below 0.");
						}
						else{
							account.setBalance(newBalance);
						}

					}

				}//end of access type -> std

			}//end of student account type

		}//end of Active account

		//if the account is not active
		else{
			System.out.println("ERROR: Not an active account.");
		}//end of not active account

		return account;
	}
	
	static public BankAccount Deposit(BankAccount account, float amount, boolean admin)
	{
		
		float depositAmount = 0.0f;
		if(amount > 0){
			depositAmount = amount;
		}
		else{
			System.out.println("ERROR: Deposit amount must be more than $0.");
		}
		
		
		//if the account is active
		if(account.isActive()){

			//non-student account
			if(!account.isStudent()){

				//admin access
				if(admin){
					float newBalance = account.getBalance() + depositAmount;
					account.setBalance(newBalance);
				}//end of acces type -> admin

				//standard access
				else{
					float newBalance = account.getBalance() + depositAmount;
					account.setBalance(newBalance);
				}//end of access type -> std

			}//end of non-student account type

			//student account
			else{
				//admin access
				if(admin){
					float newBalance = account.getBalance() + depositAmount;
					account.setBalance(newBalance);
				}//end of acces type -> admin

				//standard access
				else{
					float newBalance = account.getBalance() + depositAmount;
					account.setBalance(newBalance);
				}//end of access type -> std

			}//end of student account type

		}//end of Active account

		//if the account is not active
		else{
			System.out.println("ERROR: Not an active account.");
		}//end of not active account

		return account;
	}
	
	static public BankAccount Paybill(BankAccount account, float amount, String misc, boolean admin)
	{
		/* Other things to make sure.
		 * 	- bill amount can't bet less than 0. 
		 *  
		 * 
		 * 
		 * 1. check if account is active
		 * 		- if not active, print an error.
		 * 2. check if student or non-student
		 * 		- if nonstudent then apply a change or 0.10, and bill amount has not limit.
		 * 
		 */

		float billAmount = 0.0f;

		//first thing to check is, if the amount is greater than 0.
		if(!(amount > 0)){
			System.out.println("ERROR: amount to pay has to be greater than 0.");
		}
		else{
			billAmount = amount;
		}

		//first check if the account is active.
		if(account.isActive()){
			//account is not a student account.
			if(!account.isStudent()){
				//if an admin login
				if(admin){
					// billAmount += 0.10; //add a change of $0.10 to the billAmount.

					//new account balance. After subtracting the bill
					float newBalance = account.getBalance() - billAmount; 
					if(newBalance < 0){
						System.out.println("ERROR: transactions can not go through. + "
								+ "Account Balance results to $0.00");
					}
					else{
						//set Account balance to (newBalance)
						account.setBalance(newBalance);
					}

				}// end of if for admin login

				//if a standard login
				else{
					//bill amount has to be less than $2000.00
					if(billAmount > 2000.00f){
						System.out.println("ERROR: Bill amount is more than $2000.00");
					}
					//if the bill amount is less than $2000.00
					else{
						billAmount += 0.05; //add a change of $0.05 to the billAmount.
						//set Account balance to (current balance - billAmount)

						//new account balance. After subtracting the bill
						float newBalance = account.getBalance() - billAmount; 
						if(newBalance < 0){
							System.out.println("ERROR: transactions can not go through. + "
									+ "Account Balance results to $0.00");
						}
						else{
							//set Account balance to (newBalance)
							account.setBalance(newBalance);
						}
					}
				}// end of else for standard login
			}//end of if statement for non-student login

			//if the account is a student account.
			else{
				//if an admin login
				if(admin){
					// billAmount += 0.05; //add a change of $0.10 to the billAmount.

					//new account balance. After subtracting the bill
					float newBalance = account.getBalance() - billAmount; 
					if(newBalance < 0){
						System.out.println("ERROR: transactions can not go through. + "
								+ "Account Balance results to $0.00");
					}
					//set Account balance to (newBalance)
					account.setBalance(newBalance);
				}// end of if for admin login

				//if a standard login
				else{
					//bill amount has to be less than $2000.00
					if(billAmount > 2000.00f){
						System.out.println("ERROR: Bill amount is more than $2000.00");
					}
					//if the bill amount is less than $2000.00
					else{
						billAmount += 0.05; //add a change of $0.10 to the billAmount.
						//set Account balance to (current balance - billAmount)

						//new account balance. After subtracting the bill
						float newBalance = account.getBalance() - billAmount; 
						if(newBalance < 0){
							System.out.println("ERROR: transactions can not go through. + "
									+ "Account Balance results to $0.00");
						}
						//set Account balance to (newBalance)
						account.setBalance(newBalance);
					}
				}
			}
		}//end of isActive if statement
		else{
			System.out.println("ERROR: account is not active.");
		}
		return account;
	}
	
	static public BankAccount Changeplan(BankAccount account, String misc)
	{
		account.setStudent(!account.isStudent());
		return account;
	}
}